File: extension/test/assemble.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { assemble } from "../src/vault/assemble.ts";
import type { Note } from "../src/vault/types.ts";

/**
 * The assembled block must match the Go implementation byte-for-byte so
 * the MCP path and the extension path deliver identical context. These
 * fixtures are the reference — if you change either side, update both.
 */

function note(over: Partial<Note> & Pick<Note, "id" | "label" | "group" | "body">): Note {
  return {
    id: over.id,
    label: over.label,
    group: over.group,
    body: over.body,
    source: over.source ?? "manual",
    status: over.status ?? "active",
    created_at: over.created_at ?? "2026-07-01T12:00:00Z",
    relPath: over.relPath ?? `${over.group}/${over.label}.md`,
    ...(over.tags ? { tags: over.tags } : {}),
    ...(over.suggested_group ? { suggested_group: over.suggested_group } : {}),
  };
}

test("empty selection returns the no-memory marker", () => {
  const got = assemble({ item_ids: [] }, () => undefined);
  assert.equal(got, "<!-- pickmem: no memory selected -->\n");
});

test("empty selection under a named lens says which lens is empty", () => {
  const got = assemble({ active_lens: "Job-Hunt", item_ids: [] }, () => undefined);
  assert.equal(got, `<!-- pickmem: lens "Job-Hunt" is empty -->\n`);
});

test("single item gets a header + body, no separator", () => {
  const n = note({ id: "01A", label: "salary", group: "financial", body: "monthly base $8k" });
  const got = assemble({ item_ids: ["01A"] }, (id) => (id === "01A" ? n : undefined));
  assert.equal(got, "# salary  ·  financial\n\nmonthly base $8k\n");
});

test("multiple items get separators; header first, --- between", () => {
  const a = note({ id: "01A", label: "salary", group: "financial", body: "monthly base $8k" });
  const b = note({ id: "01B", label: "kickoff", group: "work", body: "Aug 1" });
  const got = assemble({ item_ids: ["01A", "01B"] }, (id) =>
    id === "01A" ? a : id === "01B" ? b : undefined
  );
  assert.equal(
    got,
    "# salary  ·  financial\n\nmonthly base $8k\n\n---\n\n# kickoff  ·  work\n\nAug 1\n"
  );
  // Byte-parity with Go's fmt.Fprintf(&b, "# %s  ·  %s\n\n", ...) followed by
  // TrimRight(body, "\n") + "\n".
});

test("active_lens header is present when lens is set", () => {
  const n = note({ id: "01A", label: "x", group: "g", body: "body" });
  const got = assemble({ active_lens: "Weekend", item_ids: ["01A"] }, () => n);
  assert.equal(
    got,
    "<!-- pickmem lens: Weekend -->\n\n# x  ·  g\n\nbody\n"
  );
});

test("stale ids are silently skipped, matching Go's behavior", () => {
  const a = note({ id: "01A", label: "x", group: "g", body: "body" });
  const got = assemble({ item_ids: ["01A", "01Z-deleted"] }, (id) =>
    id === "01A" ? a : undefined
  );
  // Only the live item shows up; no error, no gap.
  assert.equal(got, "# x  ·  g\n\nbody\n");
});

test("trailing newlines in body get normalized to exactly one", () => {
  const a = note({ id: "01A", label: "x", group: "g", body: "body\n\n\n\n" });
  const got = assemble({ item_ids: ["01A"] }, () => a);
  assert.equal(got, "# x  ·  g\n\nbody\n");
});

File: internal/mcp/assemble_test.go
package mcp

import (
	"testing"

	"github.com/qwgao/pickmem/internal/vault"
)

// These lock the exact byte format — mirror
// extension/test/assemble.test.ts fixture-for-fixture. If you change the
// format on either side, change both and keep the two suites in sync.

func newAssembleFixture(t *testing.T) *vault.Store {
	t.Helper()
	s, err := vault.Init(t.TempDir())
	if err != nil {
		t.Fatal(err)
	}
	return s
}

func addNote(t *testing.T, s *vault.Store, label, group, body string) *vault.Note {
	t.Helper()
	n, err := s.Add(&vault.Note{
		Frontmatter: vault.Frontmatter{Label: label, Group: group},
		Body:        body,
	})
	if err != nil {
		t.Fatal(err)
	}
	return n
}

func TestAssembleEmptySelection(t *testing.T) {
	s := newAssembleFixture(t)
	got := assemble(s, vault.Active{ItemIDs: nil})
	want := "--- pickmem: no memory selected ---\n"
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestAssembleEmptySelectionUnderLens(t *testing.T) {
	s := newAssembleFixture(t)
	got := assemble(s, vault.Active{ActiveLens: "Job-Hunt", ItemIDs: nil})
	want := "--- pickmem: lens \"Job-Hunt\" is empty ---\n"
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestAssembleSingleItem(t *testing.T) {
	s := newAssembleFixture(t)
	n := addNote(t, s, "salary", "financial", "monthly base $8k")
	got := assemble(s, vault.Active{ItemIDs: []string{n.ID}})
	want := "--- pickmem: selected memory ---\nsalary (financial): monthly base $8k\n--- end pickmem memory ---\n"
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestAssembleMultipleItemsSingleBlankLineBetween(t *testing.T) {
	s := newAssembleFixture(t)
	a := addNote(t, s, "salary", "financial", "monthly base $8k")
	b := addNote(t, s, "kickoff", "work", "Aug 1")
	got := assemble(s, vault.Active{ItemIDs: []string{a.ID, b.ID}})
	want := "--- pickmem: selected memory ---\n" +
		"salary (financial): monthly base $8k\n" +
		"\n" +
		"kickoff (work): Aug 1\n" +
		"--- end pickmem memory ---\n"
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestAssembleLensNameInHeader(t *testing.T) {
	s := newAssembleFixture(t)
	n := addNote(t, s, "x", "g", "body")
	got := assemble(s, vault.Active{ActiveLens: "Weekend", ItemIDs: []string{n.ID}})
	want := "--- pickmem: selected memory (lens: Weekend) ---\nx (g): body\n--- end pickmem memory ---\n"
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestAssembleStaleIDsSkippedSilently(t *testing.T) {
	s := newAssembleFixture(t)
	n := addNote(t, s, "x", "g", "body")
	got := assemble(s, vault.Active{ItemIDs: []string{n.ID, "01Z-deleted"}})
	want := "--- pickmem: selected memory ---\nx (g): body\n--- end pickmem memory ---\n"
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestAssembleTrailingNewlinesNormalized(t *testing.T) {
	s := newAssembleFixture(t)
	n := addNote(t, s, "x", "g", "body\n\n\n\n")
	got := assemble(s, vault.Active{ItemIDs: []string{n.ID}})
	want := "--- pickmem: selected memory ---\nx (g): body\n--- end pickmem memory ---\n"
	if got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}
